atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/TokenRangeWritesLogger.java

```diff
@@ -0,0 +1,165 @@
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableRangeMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeMap;
+import com.google.common.collect.Sets;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.logging.LoggingArgs;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.util.MathUtils;
+
+public final class TokenRangeWritesLogger {
+    private static final Logger log = LoggerFactory.getLogger(TokenRangeWritesLogger.class);
+
+    @VisibleForTesting
+    static final long THRESHOLD_WRITES_PER_TABLE = 1_000_000L;
+    @VisibleForTesting
+    static final long TIME_UNTIL_LOG_MILLIS = 24 * 60 * 60 * 1000L;
+    private static final double CONFIDENCE_FOR_LOGGING = 0.99;
+
+    private final ConcurrentMap<TableReference, TokenRangeWrites> statsPerTable = new ConcurrentHashMap<>();
+    private volatile Set<Range<LightweightOppToken>> ranges = ImmutableSet.of(Range.all());
+
+    private TokenRangeWritesLogger() {
+        // use factory method
+    }
+
+    public static TokenRangeWritesLogger createUninitialized() {
+        return new TokenRangeWritesLogger();
+    }
+
+    public void updateTokenRanges(Set<Range<LightweightOppToken>> newRanges) {
+        Preconditions.checkArgument(!newRanges.isEmpty(), "Set of ranges must not be empty!");
+        if (!Sets.symmetricDifference(ranges, newRanges).isEmpty()) {
+            ranges = newRanges;
+            statsPerTable.clear();
+        }
+    }
+
+    <V> void  markWritesForTable(Set<Map.Entry<Cell, V>> entries, TableReference tableRef) {
+        statsPerTable.putIfAbsent(tableRef, new TokenRangeWrites(tableRef, ranges));
+        TokenRangeWrites tokenRangeWrites = statsPerTable.get(tableRef);
+        entries.forEach(entry -> tokenRangeWrites.markWrite(entry.getKey()));
+        tokenRangeWrites.maybeLog();
+    }
+
+    private static final class TokenRangeWrites {
+        private final TableReference tableRef;
+        private final RangeMap<LightweightOppToken, AtomicLong> writesPerRange;
+        private final AtomicLong totalWrites = new AtomicLong(0);
+        private long lastLoggedTime = System.currentTimeMillis();
+
+        private TokenRangeWrites(TableReference tableRef, Set<Range<LightweightOppToken>> ranges) {
+            this.tableRef = tableRef;
+            ImmutableRangeMap.Builder<LightweightOppToken, AtomicLong> builder = ImmutableRangeMap.builder();
+            ranges.forEach(range -> builder.put(range, new AtomicLong(0)));
+            this.writesPerRange = builder.build();
+        }
+
+        private void markWrite(Cell cell) {
+            writesPerRange.get(LightweightOppToken.of(cell)).incrementAndGet();
+            totalWrites.incrementAndGet();
+        }
+
+        private void maybeLog() {
+            if (shouldLog()) {
```

Comment:
We check `shouldLog` here and then again in `tryLog`

Reply:
This is for thread safety -- we only enter the synchronized block when we actually want to log, but then we must check that there wasn't a race condition to avoid logging garbage.

Reply:
@jeremyk-91 Can you double check whether this makes sense before we merge?

Reply:
Think this makes sense given number of times `shouldLog` is going to be called and return `false` and the fact we don't want to log twice once the condition becomes true.